### PR TITLE
 Move default template types and template part areas to REST API

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR moves the `defaultTemplateTypes`, `defaultTemplatePartAreas`  to the `getEntityRecord( 'root', '__unstableBase' ) endpoint`. 

 Given that the impacted selectors aren't specific to the `editor` store, I decided to deprecate them all. I couldn't find a suitable location to ensure accessibility across different packages, leading to some duplicated code. Should we create a `@wordpress/template` package (similar to `@wordpress/patterns`) to house this logic? Create a dedicated package could help us with #65390 too.

## Why

This PR is necessary to remove the `core/editor` dependency across the application. In the specific, this refactor is needed to allow the package `fields` to have access to the `defaultTemplateTypes`, `defaultTemplatePartAreas` without that editor settings are initialized. More details here: https://github.com/WordPress/gutenberg/pull/65390#discussion_r1814724405

## Testing Instructions

1. Visit the Site Editor. 
2. On the sidebar click on `Pattern`.
3. Click on `All template parts`.
4. Ensure that all the patterns are visible (you should see `Header` and `Footer` labels too)
5. On the top right, click on `Add New Pattern`.
6. Create a new template part
7. Ensure that the template part is created correctly.
8. Visit the Site Editor.
9. Visit the `Blog Home`.
10. Ensure that the right panel under the template section renders the right data:
image

11. Add a block into the footer.
12. Click save.
13. Ensure that `are you ready to save` shows the right data:

image

### CODE REVIEW

File: packages/editor/src/store/selectors.js
Comment (gigitux) [inline packages/editor/src/store/selectors.js]: I think that we should deprecate this selector and create a dedicated hook to be reused across the codebase: what do you think?

File: packages/core-data/src/entities.js
Comment (youknowriad) [inline packages/core-data/src/entities.js]: Looks like these are the only fields that are camelCase? should they follow the other pattern?

**Review by youknowriad — COMMENTED** [review]
I definItely think this is the right approach to get these configuration objects from the server.

I'll defer to other about the details on naming, REST API conventions...

**Comment by youknowriad** [comment]
Do you think we can update this PR to target trunk directly (avoid all the changes related to the action)? That way we can land it first.

**Comment by gigitux** [comment]
> Do you think we can update this PR to target trunk directly (avoid all the changes related to the action)? That way we can land it first.

Thanks! Sure! Thanks for tagging other folks! After that, I will make the changes, I will mark this PR ready to be reviewed!

File: packages/editor/src/store/test/selectors.js
Comment (gigitux) [inline packages/editor/src/store/test/selectors.js:2848]: I created the same unit tests for the `getTemplateInfo` function: 2a04ba14fbe2bb575cbe53aaad63a2809c1c3408

**Comment by gigitux** [comment]
Marking this PR as ready for review. Looking forward to your feedback!

File: packages/editor/src/store/selectors.js
Comment (gigitux) [inline packages/editor/src/store/selectors.js:1735]: Technically, missing the last piece of the logic:

```ts
areas.map( ( item ) => {
				return { ...item, icon: getTemplatePartIcon( item.icon ) };
			} );
```

Not sure if I should add this line under the alternative key or simply remove it altogether.

**Comment by Mamaduka** [comment]
I don't think the [`rest_index`](https://developer.wordpress.org/reference/hooks/rest_index/) is the right place for this data. It should only be used to share minimal information about the site.

It also needs to be clarified to me **why** this change is needed. Maybe you can expand on that in the PR description or link an issue that has been discussed.

Have we considered the `WP_REST_Templates_Controller` to make this data available via REST API?

**Comment by gigitux** [comment]
> It also needs to be clarified to me **why** this change is needed. Maybe you can expand on that in the PR description or link an issue that has been discussed.

Thanks for the comment! I updated the PR description explaining why this PR is necessary. Happy to provide more context if needed.

> Have we considered the WP_REST_Templates_Controller to make this data available via REST API?

Not really. If you think that it makes sense, I could explore this path!

**Comment by youknowriad** [comment]
@Mamaduka I suggested this approach, see here https://github.com/WordPress/gutenberg/pull/65390#discussion_r1814724405 

What would be the right endpoint for you for these general filtrable non editable settings ? I can't think of better than the index personally.

**Comment by youknowriad** [comment]
For me the templates endpoint is specific to a post type, so it doesn't really make sense unless we make two API calls, one for template parts and one for templates. That would work for me but it adds more endpoints to preload, while the index endpoint is already preloaded and feels like a good place for some global uneditable settings.

**Comment by Mamaduka** [comment]
The index is available to every consumer, and that's one reason Core recommends exposing only a small amount of data about the site. I don't see how default template types and template part areas are relative in this context.

IIRC, the only reason we made Logo/Icon available was the lack of proper read capabilities for those settings.

The read-only block editor settings REST API endpoint has been mentioned several times. I can find the PRs now, but I remember the most recent discussion between @youknowriad and @TimothyBJacobs. Maybe it's time to invest in this long-term solution.

**Comment by youknowriad** [comment]
> The read-only block editor settings REST API endpoint has been mentioned several times

Yes, it's a mobile specific endpoint basically only available in the Gutenberg plugin for now. That said, what I want to avoid is an "editor settings" endpoint personally. I don't see these things as "editor" settings, they are "WordPress settings" for me.

**Comment by youknowriad** [comment]
default template and template parts are just titles and descriptions, so being available to everyone is not really an issue IMO but I can see the point.

So what would be the ideal solution knowing:

 - I don't think we should have an "editor" settings endpoint.
 - These settings are filterable in php but not editable.
 - They could be consider site settings.

Can we add these to `/wp/v2/settings` endpoint as "read only" settings?

**Comment by youknowriad** [comment]
I also want us to avoid getting stuck here, so let's think about a path forward that everyone can agree with.

**Comment by Mamaduka** [comment]
> I don't see these things as "editor" settings, they are "WordPress settings" for me.

Then, maybe we can improve the existing `/settings` endpoint to allow read-only settings and better capability management (Existing work: https://github.com/WordPress/wordpress-develop/pull/2985).

P.S. Just saw the latest comments; it seems we're on the same page :)

> I also want us to avoid getting stuck here, so let's think about a path forward that everyone can agree with.

100%.

**Comment by youknowriad** [comment]
I would love some agreement on the direction here before we actually lose time bike shedding and implementing things that won't be agreed upon. So would love input from @TimothyBJacobs and others.

The existing PR you shared @Mamaduka seems about settings that are saved to "site options". As I explained above, this is not the case for the settings that this PR is considering. These are just php filterable static values. The user doesn't update these. 

So if we are to add these to the "settings" endpoint, they can only be added adhoc if I'm not wrong. not need for generic settings registration or things like that. We could just hard code them in the endpoints no?

**Comment by youknowriad** [comment]
Ok, given the discussion on the original trac ticket didn't get any feedback. I think we can make a decision to move forward with this approach as it's a decent approach and follows the reasoning of what was done before.

**Comment by Mamaduka** [comment]
Yes, let's move forward with the current solution. I've also found a similar suggestion in an [older track ticket](https://core.trac.wordpress.org/ticket/38731#comment:1).

If a better alternative is suggested, there's more than enough time to change the approach before the next major release.

@gigitux, do you mind rebasing on top of the latest trunk and resolving merge conflicts? I'll try to do a more in-depth review later today or tomorrow.

**Comment by gigitux** [comment]
> @gigitux, do you mind rebasing on top of the latest trunk and resolving merge conflicts? I'll try to do a more in-depth review later today or tomorrow.

Done! Thanks for taking a look! 🙇

File: packages/block-library/src/template-part/edit/advanced-controls.js
Comment (youknowriad) [inline packages/block-library/src/template-part/edit/advanced-controls.js]: Seems like we might not need this temporary variable and just move the mapping into the areaOptions `map`

File: packages/editor/src/utils/get-template-info.js
Comment (youknowriad) [inline packages/editor/src/utils/get-template-info.js:18]: Ultimately, this looks like something that belongs in "core-data" but ok to keep I guess here for now.

File: packages/editor/README.md
Comment (youknowriad) [inline packages/editor/README.md]: If this is a private API, why the doc is being added here. It might be that it's exported publicly as well.

File: packages/edit-site/src/index.js
Comment (youknowriad) [inline packages/edit-site/src/index.js:98]: Good riddance.

File: packages/editor/src/store/selectors.js
Comment (youknowriad) [inline packages/editor/src/store/selectors.js:1707]: You don't know for how long I wanted to deprecate and get rid of these selectors.

File: packages/editor/src/utils/get-template-info.js
Comment (youknowriad) [reply packages/editor/src/utils/get-template-info.js:18]: Also, I'm not really sure if a big utility like that is better than smaller ones `getTemplateTitle` `getTemplateIcon` and `getTemplateDescription`. That said, feels like something to be explored separately.

File: packages/editor/src/utils/index.js
Comment (youknowriad) [inline packages/editor/src/utils/index.js]: It's funny that we already have a `getTemplatePartIcon`. It must be something that is very similar no?

**Review by youknowriad — COMMENTED** [review]
This is looking good for me, left some minor comments. I would appreciate your thoughts as well @Mamaduka

File: lib/compat/wordpress-6.8/template-parts.php
Comment (Mamaduka) [inline lib/compat/wordpress-6.8/template-parts.php]: Nit: the [`rest_index` is a filter](https://developer.wordpress.org/reference/hooks/rest_index/), let's use `add_filter` here.

We could also combine two callbacks into one, but that's optional.

File: packages/block-library/src/template-part/edit/advanced-controls.js
Comment (Mamaduka) [reply packages/block-library/src/template-part/edit/advanced-controls.js]: +1.

File: packages/edit-site/src/index.js
Comment (youknowriad) [reply packages/edit-site/src/index.js:98]: Btw, I think there's some php code that we need to remove from the core as well (setting these in the editor setting) in the backport PR

File: packages/edit-site/src/components/editor/use-editor-title.js
Comment (Mamaduka) [inline packages/edit-site/src/components/editor/use-editor-title.js]: This can be combined into a single call. The `coreStore` selectors are also extracted above; let's add `getEntityRecord` to that list. 

Bit ugly bug efficient 😄

```js
const {
	default_template_part_areas: templateAreas = [],
	default_template_types: templateTypes = [],
} = getEntityRecord( 'root', '__unstableBase' ) ?? {};
```

File: packages/edit-site/src/components/sidebar-navigation-screen-patterns/use-template-part-areas.js
Comment (Mamaduka) [inline packages/edit-site/src/components/sidebar-navigation-screen-patterns/use-template-part-areas.js:46]: This doesn't seem to be related. Does it fix a different error?

File: packages/editor/src/components/document-bar/index.js
Comment (Mamaduka) [inline packages/editor/src/components/document-bar/index.js]: Same as above: https://github.com/WordPress/gutenberg/pull/66459/files#r1843565053.

File: packages/editor/src/components/entities-saved-states/entity-record-item.js
Comment (Mamaduka) [inline packages/editor/src/components/entities-saved-states/entity-record-item.js]: Same as above: https://github.com/WordPress/gutenberg/pull/66459/files#r1843565053

File: packages/editor/src/components/entities-saved-states/entity-record-item.js
Comment (Mamaduka) [inline packages/editor/src/components/entities-saved-states/entity-record-item.js]: This new `getTemplateInfo` util requires template parts area data just to give us a template title. When working with templates, we can probably pass an empty array. What do you think?

If we decide to go in this direction, let's update all other similar occurrences.

File: packages/editor/src/components/post-card-panel/index.js
Comment (Mamaduka) [inline packages/editor/src/components/post-card-panel/index.js]: Same as above: https://github.com/WordPress/gutenberg/pull/66459/files#r1843565053

File: packages/core-data/src/entities.js
Comment (Mamaduka) [inline packages/core-data/src/entities.js:35]: Changes here also require an update to the preload path in `lib/compat/wordpress-6.8/preload.php`. See the comment above the `_fields` property.

File: packages/edit-site/src/index.js
Comment (gigitux) [reply packages/edit-site/src/index.js:98]: Yep, we can remove these lines:

https://github.com/WordPress/wordpress-develop/blob/1aa41dea3345c57cffce9059bbad728c86d1471a/src/wp-admin/site-editor.php#L64-L65

File: packages/editor/src/utils/index.js
Comment (gigitux) [reply packages/editor/src/utils/index.js]: They are the same 😁, but I don't think that we can import `getTemplatePartIcon` from `editor` package in `block-library` package.

File: packages/block-library/src/template-part/edit/advanced-controls.js
Comment (gigitux) [reply packages/block-library/src/template-part/edit/advanced-controls.js]: Addressed with 7aab231e73075333b36b4d08727c012b9ca180a3

File: packages/editor/README.md
Comment (gigitux) [reply packages/editor/README.md]: Good catch! Fixed with 1f47713e10a01561fc78f212a73bfb9f209b6461

File: lib/compat/wordpress-6.8/template-parts.php
Comment (gigitux) [reply lib/compat/wordpress-6.8/template-parts.php]: Fixed with ffdcc1d585c6879dbd07c817971480df43be65f3

File: packages/edit-site/src/components/editor/use-editor-title.js
Comment (gigitux) [reply packages/edit-site/src/components/editor/use-editor-title.js]: Fixed with e395bf67bd06b772e8057f1c78c21b78bbf2630d

File: packages/editor/src/components/document-bar/index.js
Comment (gigitux) [reply packages/editor/src/components/document-bar/index.js]: Fixed with https://github.com/WordPress/gutenberg/commit/e395bf67bd06b772e8057f1c78c21b78bbf2630d

File: packages/editor/src/components/entities-saved-states/entity-record-item.js
Comment (gigitux) [reply packages/editor/src/components/entities-saved-states/entity-record-item.js]: Fixed with https://github.com/WordPress/gutenberg/commit/e395bf67bd06b772e8057f1c78c21b78bbf2630d

File: packages/editor/src/components/post-card-panel/index.js
Comment (gigitux) [reply packages/editor/src/components/post-card-panel/index.js]: Fixed with https://github.com/WordPress/gutenberg/commit/e395bf67bd06b772e8057f1c78c21b78bbf2630d

File: packages/editor/src/utils/index.js
Comment (youknowriad) [reply packages/editor/src/utils/index.js]: yep, I think most of these belong in `core-data` but let's not focus on that for now.

File: packages/editor/src/components/entities-saved-states/entity-record-item.js
Comment (gigitux) [reply packages/editor/src/components/entities-saved-states/entity-record-item.js]: I think that you're right! I make `templateAreas` an optional parameter: 078d2f61efbc5696395558a84edb9bacae70daf8

File: lib/compat/wordpress-6.8/template-parts.php
Comment (youknowriad) [inline lib/compat/wordpress-6.8/template-parts.php]: The file name is a bit weird but it's not important. Could be rest-api.php or something.

File: packages/core-data/src/entities.js
Comment (gigitux) [reply packages/core-data/src/entities.js:35]: TIL! Thanks! fdc291d41221895d6ad6f38f8aed4fb21f9539f8

**Review by youknowriad — APPROVED** [review]

**Comment by gigitux** [comment]
Sorry for the delay, but last week I was busy with other duties! Thanks for your great review! 🙇

File: packages/edit-site/src/components/sidebar-navigation-screen-patterns/use-template-part-areas.js
Comment (gigitux) [reply packages/edit-site/src/components/sidebar-navigation-screen-patterns/use-template-part-areas.js:46]: It is related. These changes affect the loading process. Since the variables aren’t preloaded anymore, `knownAreas` might not be defined when the page loads.

File: lib/compat/wordpress-6.8/template-parts.php
Comment (Mamaduka) [reply lib/compat/wordpress-6.8/template-parts.php]: +1. I think a `rest-api.php` file is already in WP 6.8 compat dir. We can move this code there.

Comment (gigitux) [reply lib/compat/wordpress-6.8/template-parts.php]: Done! b2c0bf587e30c302d6dc0bda91881933b87e3984

**Comment by gigitux** [comment]
@Mamaduka I will wait for your approval before merging this PR!

**Review by Mamaduka — APPROVED** [review]
It looks good to merge ✅  Thanks, @gigitux!

**Comment by gigitux** [comment]
Should I handle backporting this change to WordPress?

**Comment by Mamaduka** [comment]
> Should I handle backporting this change to WordPress?

Yes. It's recommended to start backport PR as soon as possible. I think there's an even CI check for it.

See https://github.com/WordPress/gutenberg/blob/trunk/backport-changelog/readme.md for additional details.

File: packages/editor/src/store/selectors.js
Comment (oandregal) [inline packages/editor/src/store/selectors.js:1710]: shouldn't this be 6.8? (same below)

Comment (gigitux) [reply packages/editor/src/store/selectors.js:1710]: Nice catch! I'm going to open a follow-up PR! Thanks 🙇

**Comment by gigitux** [comment]
> > Should I handle backporting this change to WordPress?
> 
> Yes. It's recommended to start backport PR as soon as possible. I think there's an even CI check for it.
> 
> See https://github.com/WordPress/gutenberg/blob/trunk/backport-changelog/readme.md for additional details.

I created a backport PR: https://github.com/WordPress/wordpress-develop/pull/7895. I'm not sure if I should follow [these steps](https://github.com/WordPress/gutenberg/blob/trunk/backport-changelog/readme.md#how-to-add-a-core-backport-pr-to-the-changelog) given that this PR has already merged.

**Comment by Mamaduka** [comment]
I think those files are used to sync backport tracking issues for a release automatically. See #62973.

You could create a `.md` file in a separate PR.

**Comment by gigitux** [comment]
> I think those files are used to sync backport tracking issues for a release automatically. See #62973.
> 
> You could create a `.md` file in a separate PR.

Thanks for the feedback! I created a dedicated PR: https://github.com/WordPress/gutenberg/pull/67319

File: packages/core-data/src/entities.js
Comment (ellatrix) [reply packages/core-data/src/entities.js:35]: Why didn't you keep the same order?

Comment (ellatrix) [reply packages/core-data/src/entities.js:35]: See #67450.

**Pull request merged**
